admin/javascript/LeftAndMain.Ping.js

```diff
@@ -21,16 +28,67 @@
 			 *
 			 * This function is called by prototype when it receives notification that the user was logged out.
 			 * It uses /Security/ping for this purpose, which should return '1' if a valid user session exists.
-			 * It redirects back to the login form if the URL is either unreachable, or returns '0'.
+			 * It opens the login form in a new window if the URL is either unreachable (doesn't return HTTP 200) or doesn't return the expected '1'.
 			 */
 			_setupPinging: function() {
+				var loginWindow;	// Reference to pop-up window generated if/when ping fails.
+				var dialog;			// Reference to message dialog, if one is ever opened.
+
 				var onSessionLost = function(xmlhttp, status) {
-					if(xmlhttp.status > 400 || xmlhttp.responseText == 0) {
-						// TODO will pile up additional alerts when left unattended
-						if(window.open('Security/login')) {
-							alert('Please log in and then try again');
-						} else {
-							alert('Please enable pop-ups for this site');
+					// See if we're getting a response that contains a request to re-authenticate. If so, do that now.
+					if (xmlhttp.getResponseHeader('X-Reauthenticate')) {
+						//TODO: Change to the following line when entwine bug is found/fixed.
+						// $('.cms-container').entwine('ss').showLoginDialog();
+						containerScope.showLoginDialog();
+
+					} else if(xmlhttp.responseText !== "1") {
+						// In this case, the server may not have even responded at all. To be safe, open a new window and
+						// show a default message requesting user to login, in case the login window is already open or
+						// opens successfully. This is necessary so we don't attempt to load a login iframe to a down
+						// server (better to do that in a new window instead).
+						var message = 'Please log in and then try again.';
+
+						if (!loginWindow || (loginWindow && loginWindow.closed)) {
+							// Window hasn't yet been opened (or has been closed but must be reopened again).
+							loginWindow = window.open('Security/login');
```

Comment:
Good point, this was just an iterative improvement upon what was already setup. In fact, it took me a while to realize this (haven't touched this code in *months*), however I believe that's what this is doing right now. See `containerScope.showLoginDialog();` above. If you read the comments just below that, it only attempts to open in a new tab/window as a last resort in case the server is not responding properly. That should hopefully make it less likely for someone to be locked out of any work that they may currently have in the window which is being obscured by the login dialog and, of course, refreshing could cause loss of work.